x-pack/plugin/ml/build.gradle

```diff
@@ -74,8 +69,15 @@ dependencies {
   api "org.apache.commons:commons-math3:3.6.1"
   api "com.ibm.icu:icu4j:${versions.icu4j}"
   api "org.apache.lucene:lucene-analysis-icu:${versions.lucene}"
-  nativeBundle("org.elasticsearch.ml:ml-cpp:${project.version}@zip") {
-    changing = true
+  if (findProject(':ml-cpp')) {
+    nativeBundle project(":ml-cpp")
```

Comment:
For the case of trying to build `ml-cpp` and `elasticsearch` together it seems that the included project doesn't exist early enough in Gradle's execution for this to work.

On current `master` I can run this and it successfully builds `ml-cpp` and `elasticsearch` together:

```
./gradlew :x-pack:plugin:ml:assemble --include-build ../ml-cpp
```

However, if I run that on this PR branch it doesn't - it still downloads the dependencies.

I tried adjusting the code as follows to see if it was just `findProject` that was at fault:

```
diff --git a/x-pack/plugin/ml/build.gradle b/x-pack/plugin/ml/build.gradle
index b3d47976d9d..5e08422f41f 100644
--- a/x-pack/plugin/ml/build.gradle
+++ b/x-pack/plugin/ml/build.gradle
@@ -69,9 +69,12 @@ dependencies {
   api "org.apache.commons:commons-math3:3.6.1"
   api "com.ibm.icu:icu4j:${versions.icu4j}"
   api "org.apache.lucene:lucene-analysis-icu:${versions.lucene}"
-  if (findProject(':ml-cpp')) {
+  //if (findProject(':ml-cpp')) {
+  if (true) {
     nativeBundle project(":ml-cpp")
   } else {
     nativeBundle("org.elasticsearch.ml:ml-cpp-deps:${project.version}@zip") {
```

That failed like this _even though there was an `ml-cpp` directory in the expected place_:

```
Starting a Gradle Daemon (subsequent builds will be faster)

> Configure project :x-pack:plugin:ml
FAILURE: Build failed with an exception.

* Where:
Build file '/Users/dave/elasticsearch/x-pack/plugin/ml/build.gradle' line: 75

* What went wrong:
A problem occurred evaluating project ':x-pack:plugin:ml'.
> Project with path ':ml-cpp' could not be found in project ':x-pack:plugin:ml'.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 12s
21 actionable tasks: 21 up-to-date
```

So I am guessing that projects added by `--include-build` aren't accessible at the point where `dependencies` are configured.

Reply:
Something weird is happening with `--include-build ../ml-cpp`.

I tried making this change to the `master` branch code:

```
diff --git a/x-pack/plugin/ml/build.gradle b/x-pack/plugin/ml/build.gradle
index 495e4c10972..ed37168dc38 100644
--- a/x-pack/plugin/ml/build.gradle
+++ b/x-pack/plugin/ml/build.gradle
@@ -35,8 +35,8 @@ repositories {
 configurations {
   nativeBundle {
     resolutionStrategy.dependencySubstitution {
-      if (findProject(':ml-cpp') != null) {
-        substitute module("org.elasticsearch.ml:ml-cpp") with project(":ml-cpp")
+      if (findProject(':ml-dummy') != null) {
+        substitute module("org.elasticsearch.ml:ml-cpp") with project(":ml-dummy")
       }
     }
     resolutionStrategy.cacheChangingModulesFor 2, 'hours'
```

Then if I run `./gradlew :x-pack:plugin:ml:assemble --include-build ../ml-cpp` it still compiles the C++ code even though the substitution was changed in a way that you'd think would not do that (since `ml-dummy` doesn't exist).  So it seems that merely telling Gradle to include `../ml-cpp` would cause it to be built.

However, on this PR branch `./gradlew :x-pack:plugin:ml:assemble --include-build ../ml-cpp` never ever seems to compile the C++ code.

I am wondering if https://github.com/elastic/ml-cpp/blob/87a987393473ed232afb3903e5c590d614f0603a/gradle.properties#L5 has something to do with this.  Does that need changing?